### PR TITLE
BinaryWriter perf improvements

### PR DESCRIPTION
**Pull request by xtqqczze**

### CODE REVIEW

**Comment by xtqqczze** [comment]
@MihuBot

File: src/libraries/System.Private.CoreLib/src/System/IO/BinaryWriter.cs [hunk 1]
```diff
@@ -243,9 +243,14 @@ public virtual void Write(char[] chars, int index, int count)
...
-            BinaryPrimitives.WriteDoubleLittleEndian(buffer, value);
-            OutStream.Write(buffer);
+            if (BitConverter.IsLittleEndian)
+            {
+                WriteBytes(value);
+            }
+            else
+            {
+                WriteBytes(BinaryPrimitives.ReverseEndianness(BitConverter.DoubleToUInt64Bits(value)));
+            }
```
Comment (gfoidl) [inline src/libraries/System.Private.CoreLib/src/System/IO/BinaryWriter.cs:253]: In the changes below it's written _inline_, can this be made consistent (in either way)?

Comment (xtqqczze) [reply src/libraries/System.Private.CoreLib/src/System/IO/BinaryWriter.cs:253]: Here, we call `WriteBytes<Double>` or `WriteBytes<UInt64>`, so the _other_ methods would need to change, e.g.:

```csharp
    public virtual void Write(ulong value)
    {
        if (BitConverter.IsLittleEndian)
        {
            WriteBytes(value);
        }
        else
        {
            WriteBytes(BinaryPrimitives.ReverseEndianness(value));
        }
    }
```

**Comment by xtqqczze** [comment]
The JIT diff regressions in https://github.com/MihuBot/runtime-utils/issues/293 are entirely from additional inlining.

File: src/libraries/System.Private.CoreLib/src/System/IO/BinaryWriter.cs (hunk 1 above)
Comment (xtqqczze) [reply src/libraries/System.Private.CoreLib/src/System/IO/BinaryWriter.cs:262]: @gfoidl pushed [46ea39b](https://github.com/dotnet/runtime/pull/99775/commits/46ea39ba82b619bc395b0706ec8c38f06e3d8bac)

**Comment by xtqqczze** [comment]
@MihuBot -arm64

**Comment by xtqqczze** [comment]
@MihuBot

**Pull request closed without merging**